### ISSUE TITLE
Rule proposal: Forbid `btoa` and `atob`

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#2212]
### Description

See:

- https://developer.mozilla.org/en-US/docs/Glossary/Base64#the_unicode_problem
- https://github.com/sindresorhus/uint8array-extras/blob/75bd46dca10f7ec4fe1452fff01e1eec8c45be4d/index.js#L122-L123

In short, they don't support UTF-8

### Fail

```js
globalThis.btoa(str);
window.atob(str);
btoa(str);
atob(str)
```

### Pass

```js
import {base64ToString, stringToBase64} from "uint8array-extras";

base64ToString(str);
stringToBase64(str);
```

### Additional Info

Hopefully more native methods will be available soon: 

- https://github.com/tc39/proposal-arraybuffer-base64

**Comment by fisker** [comment]
I use `btoa(unescape(encodeURIComponent('中文')))` a lot. It works for me.

**Comment by sindresorhus** [comment]
> **unescape()**
> Deprecated: This feature is no longer recommended. Though some browsers might still support it, it may have already been removed from the relevant web standards, may be in the process of being dropped, or may only be kept for compatibility purposes. Avoid using it, and update existing code if possible; see the [compatibility table](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/unescape#browser_compatibility) at the bottom of this page to guide your decision. Be aware that this feature may cease to work at any time.
> - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/unescape

**Comment by fisker** [comment]
I know that, but it will still be in browser forever. I prefer to use it without add dependencies.

**Comment by sindresorhus** [comment]
Sure, but it's still not something we should recommend in a lint rule.

The proper solution is what `uint8array-extras` does, which is also documented here: https://developer.mozilla.org/en-US/docs/Glossary/Base64#the_unicode_problem

**Comment by silverwind** [comment]
I do agree on flagging this but it should not make any module recommendations as those are subjective. Personally I use https://www.npmjs.com/package/base-64 for example.

**Comment by fregante** [comment]
I don't know how that compares performance-wise, but that package is bigger than the whole uint8array-utils package, so maybe *it should* make recommendations.

**Comment by silverwind** [comment]
Yeah I wouln't use it today anymore. Maybe it should link to that MDN until a new standard method arises.

**Comment by silverwind** [comment]
I think https://eslint.org/docs/latest/rules/no-restricted-globals could cover at least part of this:

```json
"no-restricted-globals": [
    "error",
    {
        "name": "atob",
        "message": "Use uint8array-extras insttead"
    },
    {
        "name": "btoa",
        "message": "Use uint8array-extras instead"
    }
]
```

I think it will definitely match on `atob`, but likely not on `window.atob` or `globalThis.atob`, but I would consider that a rule bug.

**Comment by sindresorhus** [comment]
> Assuming eslint core will stubbornly refuse to acknowledge it as a bug

Open an ESLint issue. I don't see why they would refuse that.

**Comment by silverwind** [comment]
I won't, had too many bad experiences with opening issues at eslint because they always refuse any such behaviour changes in their rules. At this point, I think it's better to fork and fix [no-restricted-globals](https://eslint.org/docs/latest/rules/no-restricted-globals). [This playground](https://eslint.org/play/#eyJ0ZXh0IjoiYXRvYigpXG53aW5kb3cuYXRvYigpXG5nbG9iYWxUaGlzLmF0b2IoKSIsIm9wdGlvbnMiOnsicnVsZXMiOnsibm8tcmVzdHJpY3RlZC1nbG9iYWxzIjpbImVycm9yIiwiYXRvYiJdfSwibGFuZ3VhZ2VPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=) confirms above suspicion.

**Comment by JounQin** [comment]
I'm using [`ab64`](https://github.com/un-ts/ab64) personally.

**Comment by JounQin** [comment]
Why recommending `uint8array-extras` is acceptable but another library is not?

**Comment by lishaduck** [comment]
...and https://github.com/eslint/eslint/pull/19939 has been merged...

I do think there's value in a simple way to discover and enable these sorts of rules, but that's far out-of-scope for unicorn.
